## Imports

In [0]:
from pyspark.sql.functions import *
from pyspark.sql.types import *

## Reading the data

In [0]:
inventory = spark.read.parquet('dbfs:/mnt/shopping_data_dbfs/Gold-layer/BG-Thushara/capstone-azure-training/raw/main/Data/New_data/inventory_data.parquet')

In [0]:
sales = spark.read.parquet('dbfs:/mnt/shopping_data_dbfs/Gold-layer/BG-Thushara/capstone-azure-training/raw/main/Data/New_data/sales_data.parquet')

In [0]:

product = spark.read.parquet('dbfs:/mnt/shopping_data_dbfs/Gold-layer/BG-Thushara/capstone-azure-training/raw/main/Data/New_data/product_data.parquet')

In [0]:
product_lev_sales = spark.read.parquet('dbfs:/mnt/shopping_data_dbfs/Gold-layer/BG-Thushara/capstone-azure-training/raw/main/Data/New_data/product_level_sales_data.parquet')

## Displaying glimpse of the data.

In [0]:
inventory.display()

Date Product_ID Inventory 2019-01-03 PROD039 158 2019-01-12 PROD019 80 2019-01-14 PROD007 73 2019-01-22 PROD036 352 2019-02-01 PROD005 872 2019-02-01 PROD016 423 2019-02-07 PROD035 403 2019-02-21 PROD039 456 2019-02-27 PROD002 438 2019-02-28 PROD033 502 2019-03-06 PROD022 509 2019-03-08 PROD029 239 2019-03-12 PROD009 799 2019-03-15 PROD006 111 2019-03-23 PROD035 498 2019-04-01 PROD020 222 2019-04-02 PROD014 188 2019-04-06 PROD023 109 2019-04-15 PROD020 975 2019-04-16 PROD015 995 2019-04-16 PROD024 777 2019-04-16 PROD029 986 2019-04-17 PROD022 616 2019-04-21 PROD030 664 2019-04-22 PROD021 798 2019-05-01 PROD008 812 2019-05-02 PROD012 60 2019-05-06 PROD026 705 2019-05-17 PROD011 694 2019-05-23 PROD024 100 2019-05-25 PROD034 100 2019-05-26 PROD005 644 2019-05-26 PROD040 910 2019-06-08 PROD004 333 2019-06-09 PROD001 243 2019-06-14 PROD020 306 2019-06-16 PROD027 638 2019-06-20 PROD032 384 2019-06-20 PROD038 561 2019-06-25 PROD039 808 2019-06-28 PROD031 296 2019-07-01 PROD028 964 2019-07-25 PROD018 452 2019-07-29 PROD024 278 2019-08-02 PROD028 910 2019-08-05 PROD030 301 2019-08-11 PROD009 442 2019-08-13 PROD029 140 2019-09-01 PROD019 536 2019-09-10 PROD018 254 2019-09-24 PROD002 518 2019-10-03 PROD036 489 2019-10-07 PROD023 915 2019-10-09 PROD023 591 2019-10-15 PROD035 351 2019-10-20 PROD009 624 2019-10-26 PROD037 199 2019-10-29 PROD006 859 2019-10-30 PROD014 397 2019-11-05 PROD017 349 2019-11-05 PROD039 700 2019-11-07 PROD001 317 2019-11-16 PROD005 308 2019-11-29 PROD039 831 2019-12-09 PROD014 827 2019-12-09 PROD025 75 2019-12-12 PROD002 572 2019-12-21 PROD038 110 2019-12-21 PROD039 179 2019-12-23 PROD027 410 2020-01-06 PROD017 682 2020-01-10 PROD031 582 2020-01-11 PROD011 149 2020-01-14 PROD002 345 2020-01-17 PROD038 432 2020-01-19 PROD010 266 2020-01-25 PROD032 948 2020-01-28 PROD004 373 2020-01-28 PROD029 225 2020-02-07 PROD025 342 2020-02-19 PROD016 680 2020-02-27 PROD019 533 2020-02-28 PROD022 398 2020-02-29 PROD024 590 2020-03-02 PROD032 128 2020-03-04 PROD025 150 2020-03-06 PROD013 482 2020-03-11 PROD019 856 2020-03-12 PROD004 941 2020-03-18 PROD039 891 2020-03-26 PROD014 121 2020-04-07 PROD011 102 2020-04-16 PROD001 230 2020-04-16 PROD019 340 2020-04-23 PROD017 907 2020-04-24 PROD001 794 2020-04-24 PROD014 619 2020-05-09 PROD013 69 2020-05-11 PROD024 903 2020-05-20 PROD040 301 2020-05-21 PROD003 751 2020-05-22 PROD003 981 2020-05-22 PROD006 737 2020-05-22 PROD040 177 2020-05-24 PROD012 929 2020-05-26 PROD016 951 2020-05-26 PROD027 941 2020-06-05 PROD013 809 2020-06-09 PROD006 438 2020-06-11 PROD014 956 2020-06-12 PROD022 996 2020-06-15 PROD028 591 2020-06-25 PROD019 654 2020-07-02 PROD023 153 2020-07-05 PROD009 434 2020-07-08 PROD040 298 2020-07-09 PROD032 109 2020-07-18 PROD006 371 2020-07-28 PROD001 529 2020-08-04 PROD033 367 2020-08-10 PROD026 874 2020-08-14 PROD017 960 2020-08-15 PROD002 385 2020-08-27 PROD017 212 2020-08-27 PROD020 74 2020-08-31 PROD023 243 2020-09-09 PROD021 111 2020-09-11 PROD017 905 2020-09-22 PROD003 172 2020-09-29 PROD013 874 2020-09-29 PROD015 522 2020-10-03 PROD023 168 2020-10-04 PROD038 966 2020-10-15 PROD012 242 2020-10-17 PROD020 84 2020-10-18 PROD020 799 2020-11-09 PROD037 490 2020-11-12 PROD037 507 2020-11-14 PROD008 889 2020-11-16 PROD025 461 2020-11-17 PROD012 864 2020-11-20 PROD013 303 2020-11-28 PROD029 871 2020-12-02 PROD038 551 2020-12-04 PROD024 97 2020-12-05 PROD037 560 2020-12-12 PROD037 278 2020-12-15 PROD038 261 2020-12-26 PROD001 815 2020-12-26 PROD002 1000 2020-12-27 PROD007 343 2020-12-28 PROD033 688 2020-12-31 PROD028 215 2021-01-01 PROD034 839 2021-01-03 PROD020 355 2021-01-12 PROD010 294 2021-01-12 PROD040 82 2021-01-18 PROD023 200 2021-01-19 PROD027 399 2021-01-31 PROD013 889 2021-02-09 PROD009 553 2021-02-25 PROD019 781 2021-02-27 PROD006 888 2021-02-28 PROD032 875 2021-03-02 PROD038 966 2021-03-06 PROD035 518 2021-03-08 PROD024 623 2021-03-09 PROD025 393 2021-03-12 PROD023 662 2021-03-24 PROD002 308 2021-03-25 PROD029 693 2021-03-26 PROD007 725 2021-04-01 PROD034 549 202

In [0]:
sales.display()

sale_id store_id customer_id product_id sale_date sale_amount SALE0470 45 2249 PROD008 2020-04-14 746 SALE0026 44 1894 PROD035 2019-01-26 746 SALE0458 51 1049 PROD026 2020-04-02 746 SALE0705 27 1098 PROD011 2020-12-05 746 SALE0816 57 2105 PROD033 2021-03-26 746 SALE0301 49 2188 PROD009 2019-10-28 746 SALE0511 44 2060 PROD020 2020-05-25 746 SALE0016 33 1726 PROD037 2019-01-16 746 SALE0598 2 2203 PROD033 2020-08-20 746 SALE0232 52 2585 PROD012 2019-08-20 746 SALE0407 8 2161 PROD022 2020-02-11 746 SALE0622 49 1966 PROD031 2020-09-13 746 SALE0382 35 1404 PROD019 2020-01-17 746 SALE0392 51 1828 PROD022 2020-01-27 746 SALE0445 46 1180 PROD002 2020-03-20 746 SALE0365 5 2313 PROD027 2019-12-31 746 SALE0850 27 1169 PROD011 2021-04-29 746 SALE0535 3 2231 PROD032 2020-06-18 746 SALE0918 19 1523 PROD035 2021-07-06 746 SALE0246 2 1810 PROD018 2019-09-03 746 SALE0331 54 1259 PROD021 2019-11-27 746 SALE0732 59 2648 PROD018 2021-01-01 746 SALE0257 3 1320 PROD006 2019-09-14 746 SALE0565 40 2968 PROD007 2020-07-18 746 SALE0665 37 2693 PROD013 2020-10-26 746 SALE0417 28 1866 PROD018 2020-02-21 746 SALE0984 25 2869 PROD006 2021-09-10 746 SALE0488 48 2616 PROD024 2020-05-02 746 SALE0544 14 2065 PROD020 2020-06-27 746 SALE0306 21 2945 PROD020 2019-11-02 746 SALE0141 46 1367 PROD018 2019-05-21 746 SALE0524 43 1057 PROD024 2020-06-07 746 SALE0275 35 2292 PROD033 2019-10-02 746 SALE0502 31 2067 PROD015 2020-05-16 746 SALE0244 35 2238 PROD004 2019-09-01 746 SALE0060 30 1229 PROD033 2019-03-01 746 SALE0180 33 2581 PROD034 2019-06-29 1261 SALE0313 8 2064 PROD026 2019-11-09 517 SALE0707 54 2313 PROD019 2020-12-07 117 SALE0714 43 1087 PROD010 2020-12-14 1216 SALE0758 58 1385 PROD031 2021-01-27 141 SALE0071 25 1330 PROD037 2019-03-12 531 SALE0147 55 2154 PROD039 2019-05-27 583 SALE0305 4 1485 PROD025 2019-11-01 1029 SALE0730 34 1820 PROD029 2020-12-30 266 SALE0744 21 2930 PROD002 2021-01-13 978 SALE0111 52 1621 PROD035 2019-04-21 795 SALE0409 59 2997 PROD027 2020-02-13 714 SALE0461 19 2234 PROD027 2020-04-05 1397 SALE0550 58 1327 PROD005 2020-07-03 298 SALE0666 37 1184 PROD022 2020-10-27 277 SALE0700 9 1703 PROD008 2020-11-30 595 SALE0831 47 1642 PROD027 2021-04-10 1409 SALE0985 29 2225 PROD040 2021-09-11 695 SALE0233 30 2406 PROD001 2019-08-21 573 SALE0505 46 1550 PROD013 2020-05-19 560 SALE0766 26 2784 PROD009 2021-02-04 394 SALE0485 40 1740 PROD022 2020-04-29 299 SALE0520 31 2119 PROD033 2020-06-03 1445 SALE0146 56 2323 PROD003 2019-05-26 77 SALE0151 44 2450 PROD023 2019-05-31 1065 SALE0194 40 2723 PROD008 2019-07-13 65 SALE0551 48 2570 PROD004 2020-07-04 877 SALE0801 42 1507 PROD038 2021-03-11 239 SALE0863 2 2612 PROD006 2021-05-12 423 SALE0889 57 2346 PROD033 2021-06-07 1168 SALE0570 21 1042 PROD032 2020-07-23 528 SALE0701 19 2650 PROD002 2020-12-01 102 SALE0212 51 1846 PROD025 2019-07-31 940 SALE0239 41 2763 PROD037 2019-08-27 1167 SALE0369 2 1739 PROD018 2020-01-04 273 SALE0581 2 1519 PROD026 2020-08-03 716 SALE0603 31 2270 PROD029 2020-08-25 839 SALE0031 4 1954 PROD009 2019-01-31 1103 SALE0573 27 1829 PROD001 2020-07-26 88 SALE0677 35 2380 PROD035 2020-11-07 596 SALE0523 57 1843 PROD028 2020-06-06 1441 SALE0913 40 1511 PROD007 2021-07-01 861 SALE0969 42 1737 PROD031 2021-08-26 872 SALE0183 9 2169 PROD002 2019-07-02 1424 SALE0487 6 1010 PROD027 2020-05-01 1262 SALE0631 43 2444 PROD033 2020-09-22 846 SALE0851 9 2840 PROD017 2021-04-30 873 SALE0995 40 1261 PROD031 2021-09-21 607 SALE0367 57 1636 PROD029 2020-01-02 846 SALE0651 23 1026 PROD036 2020-10-12 203 SALE0683 35 1600 PROD009 2020-11-13 876 SALE1000 11 1356 PROD035 2021-09-26 77 SALE0203 34 1568 PROD034 2019-07-22 702 SALE0410 28 1381 PROD018 2020-02-14 1224 SALE0545 39 1145 PROD026 2020-06-28 1371 SALE0812 46 2568 PROD029 2021-03-22 136 SALE0911 18 1232 PROD007 2021-06-29 717 SALE0219 14 1182 PROD015 2019-08-07 79 SALE0348 40 2353 PROD010 2019-12-14 822 SALE0847 46 2199 PROD036 2021-04-26 257 SALE0038 36 1042 PROD005 2019-02-07 849 SALE0056 12 1656 PROD021 2019-02-25 542 SALE0090 41 1962 PROD038

In [0]:
product.display()

Product_ID Inventory sale_amount frq_of_product PROD011 528231 24756 34 PROD035 534614 17344 27 PROD003 529347 33244 39 PROD025 524151 12971 17 PROD034 501157 12517 15 PROD023 524786 16214 22 PROD036 525849 14863 22 PROD026 530282 19283 24 PROD037 526487 20564 31 PROD027 529457 20049 28 PROD029 533065 27487 34 PROD018 526332 16323 24 PROD040 530252 23901 26 PROD024 526683 23518 29 PROD009 528149 25104 32 PROD028 528768 16954 22 PROD016 539375 15200 20 PROD020 506627 13088 20 PROD014 540554 14144 21 PROD038 514133 14928 25 PROD015 531994 12622 18 PROD012 513749 22531 29 PROD006 526510 19671 28 PROD007 526614 19283 22 PROD010 522456 22953 34 PROD004 528754 17049 23 PROD005 519278 20974 27 PROD008 526004 14254 21 PROD017 513305 20831 25 PROD030 529408 18441 23 PROD033 529081 25528 27 PROD021 516925 15077 22 PROD039 525764 18568 21 PROD022 539531 19306 26 PROD019 536987 20530 27 PROD031 531260 15193 23 PROD013 534586 19076 23 PROD002 522988 13614 21 PROD032 530677 18264 27 PROD001 510288 11510 22

In [0]:
product_lev_sales.display()

Product_ID Date sale_amount Inventory PROD002 2019-07-30 199 869 PROD040 2019-07-19 1282 675 PROD025 2020-04-27 1112 930 PROD020 2020-05-18 574 746 PROD003 2021-02-23 604 635 PROD026 2020-05-11 1189 991 PROD005 2020-09-15 481 476 PROD021 2019-10-03 228 97 PROD026 2020-01-13 582 595 PROD034 2020-10-29 1030 819 PROD022 2019-06-05 798 829 PROD035 2019-07-06 378 892 PROD001 2021-08-27 674 638 PROD002 2021-08-02 454 195 PROD017 2020-03-26 711 980 PROD030 2020-11-24 1169 739 PROD018 2019-08-08 840 879 PROD009 2019-04-19 666 180 PROD008 2019-05-29 584 892 PROD013 2021-05-03 276 795 PROD006 2019-11-04 64 926 PROD038 2019-10-04 1374 427 PROD010 2020-09-09 905 247 PROD036 2021-06-30 536 933 PROD027 2021-04-10 1409 687 PROD038 2020-02-29 157 335 PROD010 2020-07-20 238 773 PROD018 2019-02-24 1313 800 PROD030 2021-05-27 947 511 PROD009 2019-08-19 776 990 PROD005 2020-01-03 397 533 PROD001 2019-12-15 918 929 PROD013 2020-01-28 534 803 PROD001 2020-07-26 88 541 PROD011 2019-08-23 504 824 PROD001 2020-03-14 1067 736 PROD003 2021-01-14 736 902 PROD018 2019-10-19 523 353 PROD027 2019-12-31 746 211 PROD002 2020-03-20 746 600 PROD028 2019-02-12 487 115 PROD012 2021-08-30 1457 765 PROD020 2019-09-09 1018 282 PROD007 2021-01-07 950 56 PROD031 2020-10-04 500 769 PROD009 2019-09-02 326 368 PROD005 2021-09-18 1258 900 PROD026 2021-02-20 1370 712 PROD024 2020-06-07 746 399 PROD036 2019-02-09 1121 982 PROD024 2019-12-28 1437 64 PROD025 2019-02-10 1196 818 PROD003 2020-02-18 1184 529 PROD036 2019-12-06 754 787 PROD033 2020-08-22 1466 514 PROD016 2019-05-11 209 220 PROD025 2019-07-31 940 950 PROD036 2020-10-12 203 897 PROD029 2021-09-01 611 495 PROD006 2020-06-24 1275 81 PROD003 2020-06-25 1383 917 PROD029 2020-01-02 846 474 PROD019 2020-01-22 1113 884 PROD037 2019-03-05 1096 301 PROD028 2021-04-15 91 503 PROD013 2020-12-12 429 867 PROD038 2021-06-04 524 768 PROD010 2021-03-21 1204 560 PROD037 2020-02-02 463 200 PROD032 2020-12-31 564 55 PROD010 2021-08-13 1080 612 PROD014 2019-06-27 499 871 PROD003 2021-09-27 1036 958 PROD026 2020-02-17 534 758 PROD024 2019-11-21 296 482 PROD011 2019-03-26 416 558 PROD007 2021-07-01 861 171 PROD026 2019-12-23 1169 260 PROD005 2019-04-10 287 891 PROD006 2019-02-14 223 473 PROD010 2020-11-27 592 640 PROD009 2019-10-28 746 703 PROD036 2020-10-07 23 887 PROD009 2020-11-13 876 402 PROD004 2019-08-02 227 423 PROD031 2021-01-15 650 265 PROD035 2021-07-11 1261 606 PROD016 2019-04-20 262 533 PROD026 2020-08-03 716 258 PROD003 2021-07-03 1165 647 PROD016 2019-05-17 872 819 PROD002 2021-09-08 583 272 PROD010 2020-01-19 null 266 PROD021 2020-09-09 null 111 PROD003 2020-08-03 null 538 PROD015 2020-08-19 null 540 PROD039 2020-12-30 null 663 PROD036 2021-06-23 null 292 PROD021 2019-01-07 null 574 PROD037 2019-11-23 null 518 PROD001 2019-06-02 null 640 PROD039 2021-03-07 null 689 PROD015 2019-06-20 null 162 PROD038 2020-01-02 null 531 PROD007 2020-09-27 null 303 PROD005 2021-03-10 null 129 PROD018 2021-06-06 null 236 PROD040 2019-12-15 null 568 PROD035 2019-03-27 null 292 PROD021 2019-08-22 null 797 PROD024 2019-03-07 null 593 PROD013 2019-03-17 null 942 PROD002 2019-02-25 null 377 PROD008 2021-09-10 null 224 PROD025 2020-02-27 null 642 PROD022 2020-01-29 null 316 PROD019 2019-11-14 null 704 PROD023 2019-04-24 null 296 PROD028 2019-11-21 null 802 PROD029 2020-07-25 null 636 PROD035 2021-09-27 null 784 PROD021 2019-10-01 null 795 PROD027 2021-09-01 null 651 PROD027 2020-08-11 null 808 PROD009 2019-05-08 null 978 PROD031 2021-01-19 null 183 PROD032 2021-04-27 null 430 PROD003 2021-05-23 null 968 PROD035 2019-07-29 null 852 PROD028 2019-12-03 null 254 PROD033 2019-01-23 null 73 PROD002 2021-09-26 null 99 PROD007 2020-05-16 null 842 PROD029 2019-02-14 null 97 PROD013 2021-04-03 null 535 PROD006 2019-05-02 null 289 PROD008 2019-12-12 null 421 PROD036 2020-06-11 null 493 PROD011 2021-05-17 null 643 PROD027 2019-10-05 null 228 PROD021 2021-05-29 null 801 PROD033 2019-10-31 null 198 PROD030 2021-04-26 null 756 PROD005 2020-02-01 null 703 PROD01

## Insights and Visualization

In [0]:
product.where(col('frq_of_product') == product.select(max(col('frq_of_product'))).collect()[0][0]).select('Product_ID','frq_of_product').show()

+----------+--------------+
|Product_ID|frq_of_product|
+----------+--------------+
|   PROD003|            39|
+----------+--------------+



- Therfore, **PROD003** is product on demand, or most sold product.

In [0]:
sales.groupBy('customer_id').agg(sum('sale_amount').alias('sale_amount')).select(max(col('sale_amount'))).display()

max(sale_amount) 3599

In [0]:
sales.groupBy('customer_id').agg(sum('sale_amount').alias('sale_amount')).filter(col('sale_amount')==3599).display()

customer_id sale_amount 2584 3599

- The customer posessing **2584 - customer id** is the most loyal customer in terms of sales.

In [0]:
sales.groupBy('customer_id') \
     .agg(sum('sale_amount').alias('sale_amount')) \
     .orderBy(col('sale_amount').desc()) \
     .limit(10).display()

customer_id sale_amount 2584 3599 1839 3558 1381 3532 2450 3527 1621 3475 1402 3302 1671 3261 1391 3184 1010 3128 2570 2972

Databricks visualization. Run in Databricks to view.

In [0]:
sales.groupBy('store_id') \
     .agg(sum('sale_amount').alias('sale_amount')) \
     .orderBy(col('sale_amount').desc()) \
     .limit(5).display()

store_id sale_amount 54 19093 41 18988 9 18659 50 18643 52 18349

Databricks visualization. Run in Databricks to view.

In [0]:
sales.groupBy('sale_date').agg(sum('sale_amount').alias('sale_amount')).select(max(col('sale_amount'))).display()

max(sale_amount) 1493

In [0]:
sales.groupBy('sale_date').agg(sum('sale_amount').alias('sale_amount')).filter(col('sale_amount')==1493).display()

sale_date sale_amount 2021-05-29 1493

- on a whole, 2021-05-29 is the day, when the best sales happened.

In [0]:
inventory.groupBy('Product_ID').agg(sum('Inventory').alias('Inventory')).orderBy(desc(col('Inventory'))).display()

Product_ID Inventory PROD014 540554 PROD022 539531 PROD016 539375 PROD019 536987 PROD035 534614 PROD013 534586 PROD029 533065 PROD015 531994 PROD031 531260 PROD032 530677 PROD026 530282 PROD040 530252 PROD027 529457 PROD030 529408 PROD003 529347 PROD033 529081 PROD028 528768 PROD004 528754 PROD011 528231 PROD009 528149 PROD024 526683 PROD007 526614 PROD006 526510 PROD037 526487 PROD018 526332 PROD008 526004 PROD036 525849 PROD039 525764 PROD023 524786 PROD025 524151 PROD002 522988 PROD010 522456 PROD005 519278 PROD021 516925 PROD038 514133 PROD012 513749 PROD017 513305 PROD001 510288 PROD020 506627 PROD034 501157

- **PROD014** is the most available product.